[Concurrency] Stop TaskGroup from holding onto Tasks forever (leaking)

ktoso opened the pull request:
Just to run the tests on CI for now.

rdar://81130259
https://bugs.swift.org/browse/SR-14973

ktoso committed:
[Concurrency] Fix group child tasks not being released

The proper handling of task group child tasks is that:
- if it completes a waiting task immediately, we don't need to retain it
  - we just move the value to the waiting task and can destroy the task
- if we need to store the ready task and wait for a waiting task (for a
  task that hits `await group.next()`) then we need to retain the ready
  task.
  - as the waiting task arrives, we move the value from the ready task
    to the waiting task, and swift_release the ready task -- it will now
    be destroyed safely.

ktoso commented:
@swift-ci please smoke test Linux

ktoso commented:
@swift-ci please smoke test macOS

ktoso commented on stdlib/public/Concurrency/TaskGroup.cpp:
```diff
@@ -623,10 +633,12 @@ void TaskGroupImpl::offer(AsyncTask *completedTask, AsyncContext *context) {
   // it must remain alive until the task group is alive.
-  swift_retain(completedTask);
+//  SWIFT_TASK_DEBUG_LOG(" enqueue completed, retain completed task = %p", completedTask);
+//  swift_retain(completedTask);
```
we basically skip doing any of the manual retains since they are not necessary ever since `fillGroupNextResult` became a thing I believe.

Previously we stored the entire tasks, so the dance was necessary, now the "fill in" moves ownership of the value out of the task

ktoso commented on test/Concurrency/Runtime/async_taskgroup_dontLeakTasks.swift:
```diff
@@ -0,0 +1,43 @@
+@main struct Main {
+  static func main() async {
+    await test_taskGroup_next()
+  }
```
no great way to test this without having task `destroy` print... is there some trick we could pull here?

al45tair replied:
You could put the tasks into an array of weak references, then check that they're all nil after you're done?

ktoso replied:
Oh, let me try that!

ktoso replied:
Hmm... it's pretty tricky, Task isn't really just a class, but we'd have to use the Builtin.NativeObject which we can't here... but trying to pretend it is AnyObject gets me into compiler crashes 😢  

There is no way to just "get a task" as a class; there is only `withUnsafeCurrentTask` that gives us a struct wrapper around the native object. I'm not sure it'd even work (I'll try a little bit more though)

ktoso commented:
@swift-ci please smoke test

ogres commented on test/Concurrency/Runtime/async_taskgroup_dontLeakTasks.swift:
```diff
@@ -0,0 +1,43 @@
+  }
+
+  // CHECK: result with group.next(): 100
+  print("result with group.next(): \(100)")
```
should this use result from task-group instead of hardcoded 100?

ktoso replied:
Sure yeah; thanks

ktoso committed:
[Concurrency] TaskGroup children remove their records as they complete

If we didn't do this (and we didn't), the tasks get released as we
perform the next() impl, and move the value from the ready task to the
waiting task. Then, the ready task gets destroyed.

But as the task group exists, it performs a cancelAll() and that
iterates over all records. Those records were not removed previously
(!!!) which meant we were pointing at now deallocated tasks.

Previously this worked because we didn't deallocate the tasks, so they
leaked, but we didn't crash. With the memory leak fixed, this began to
crash since we'd attempt to cancel already destroyed tasks.

Solution:
- Remove task records whenever they complete a waiting task.
- This can ONLY be done by the "group owning task" itself, becuause
  the contract of ONLY this task being allowed to modify records. o
  It MUST NOT be done by the completing tasks as they complete, as it
  would race with the owning task modifying this linked list of child
  tasks in the group record.

ktoso commented:
@swift-ci please smoke test

ktoso commented on include/swift/ABI/TaskStatus.h:
```diff
@@ -211,6 +210,33 @@ class TaskGroupTaskStatusRecord : public TaskStatusRecord {
     cur->childFragment()->setNextChild(child);
   }
 
+  void detachChild(AsyncTask *child) {
```
Feels pretty silly, we never removed the tasks from the group record.

This must only be done by the group and the task running the group because we don't take locks here -- _on purpose_ and by design, only the parent task is allowed to access and mutate these things.

ktoso replied:
Things worked without the removal because the tasks were leaking and never being deallocated, so as the group existed it'd hit cancelAll() which iterates over this record to cancel tasks. This was fine then since the tasks existed still.

As we fixed the bug about the child tasks lifecycle, and being properly released when they _fulfill_ a waiting tasks's value by copying it over, and then are destroyed... if we did a cancelAll() after a task had completed (and been destroyed), we'd iterate here and call cancel on trash memory.

So this PR fixes both: the leak, and the proper cancellation record handling

ktoso commented on stdlib/public/Concurrency/TaskGroup.cpp:
```diff
@@ -490,16 +494,8 @@ void TaskGroupImpl::destroy() {
-    swift_release(item.getTask());
-    taskDequeued = readyQueue.dequeue(item);
-  }
-  mutex.unlock(); // TODO: remove fragment lock, and use status for synchronization
```
this was not going to do anything because we handle this in the swift side of the impl

ktoso commented on stdlib/public/Concurrency/TaskGroup.cpp:
```diff
@@ -703,13 +701,22 @@ static void swift_taskGroup_wait_next_throwingImpl(
+      // it would be null for PollStatus::Empty, then we don't need to release
+      group->detachChild(polled.retainedTask);
+      swift_release(polled.retainedTask);
+    }
```
the release must of course happen AFTER the detach since the detach will inspect the task for it's next task in the invasive linked list that they form.

ktoso commented on stdlib/public/Concurrency/TaskPrivate.h:
```diff
@@ -42,9 +42,12 @@
-  fprintf(stderr, "[%lu] " fmt "\n", (unsigned long)_swift_get_thread_id(),    \
+  fprintf(stderr, "[%lu] [%s:%d](%s) " fmt "\n",                               \
+          (unsigned long)_swift_get_thread_id(),                               \
+          __FILE__, __LINE__, __FUNCTION__,                                    \
```
I hope no-one minds making this a bit more informative? It is a life-saver at least for me to have this include where the log is coming from... 🙏

mikeash replied:
Great idea.

ktoso replied:
Yay!

ktoso commented on test/Concurrency/Runtime/async_task_withUnsafeCurrentTask.swift:
```diff
@@ -0,0 +1,39 @@
+// RUN: %target-run-simple-swift( -Xfrontend -disable-availability-checking -parse-as-library) 2>&1 | %FileCheck %s --dump-input=always
+// REQUIRES: executable_test
+// REQUIRES: concurrency
+// REQUIRES: swift_task_debug_log
```
As I did not find a better way to test it... this depends on the SWIFT_TASK_DEBUG_LOG prints from ~AsyncTask

ktoso commented:
@swift-ci please smoke test

DougGregor commented on stdlib/public/Concurrency/TaskGroup.cpp:
```diff
@@ -839,13 +850,16 @@ static void swift_taskGroup_cancelAllImpl(TaskGroup *group) {
     return false;
   }
 
+  // FIXME: must also remove the records!!!!
```
Is this still TODO?

ktoso replied:
Uff I think so... sorry, ran out of weekend :-\ 

I'm continuing to look at this and the other reported leak...

DougGregor reviewed (approved):
Looks good!
